pkg/liquidity-mining/test/DistributionScheduler.test.ts

```diff
@@ -0,0 +1,394 @@
+          "DistributionScheduler is not reward token's distributor"
+        );
+      });
+    });
+
+    context('when DistributionScheduler is the distributor for reward token', () => {
+      sharedBeforeEach('set reward token distributor to DistributionScheduler', async () => {
+        await rewardTokenDistributor.add_reward(rewardToken.address, distributionScheduler.address);
+      });
+
+      context('when distribution is scheduled in the past', () => {
+        it('reverts', async () => {
+          await expect(scheduleDistribution(1, 0)).to.be.revertedWith(
+            'Distribution can only be scheduled for the future'
+          );
+        });
+      });
+
+      context('when distribution is scheduled in the future', () => {
+        let startTime: BigNumber;
+
+        sharedBeforeEach('set the first valid timestamp to schedule rewards', async () => {
+          startTime = roundUpTimestamp(await currentTimestamp());
+        });
+
+        context('when distribution does not start at the beginning of a week', () => {
+          it('reverts', async () => {
+            const invalidTimestamp = startTime.add(1);
+
+            await expect(
+              distributionScheduler.scheduleDistribution(
+                rewardTokenDistributor.address,
+                rewardToken.address,
+                1,
+                invalidTimestamp
+              )
+            ).to.be.revertedWith('Distribution must start at the beginning of the week');
+          });
+        });
+
+        context('when distribution starts at the beginning of a week', () => {
+          context('no rewards currently exist for this gauge', () => {
+            it('updates the the head node to point at the new node', async () => {
+              expect(await getNextNodeKey(HEAD)).to.be.eq(0);
+
+              await scheduleDistribution(1, startTime);
+
+              expect(await getNextNodeKey(HEAD)).to.be.eq(startTime);
+            });
+
+            it('writes the expected node', async () => {
+              const oldRewardNode = await getRewardNode(startTime);
+              expect(oldRewardNode.amount).to.be.eq(0);
+
+              await scheduleDistribution(1, startTime);
+
+              const newRewardNode = await getRewardNode(startTime);
+              expect(newRewardNode.amount).to.be.eq(1);
+            });
+          });
+
+          context('when there are already rewards scheduled for this gauge', () => {
+            sharedBeforeEach('schedule distributions', async () => {
+              for (let i = 0; i < 10; i++) {
+                await scheduleDistribution(100, startTime.add(i * 2 * WEEK));
+              }
+            });
+
+            context('when a reward is being added after the last existing reward', () => {
+              let insertedTime: BigNumber;
+
+              sharedBeforeEach('set insertedTime', async () => {
+                insertedTime = startTime.add(20 * WEEK);
```

Comment:
```suggestion
                insertedTime = startTime.add(20 * WEEK);
```

Same here, we're _right_ at the edge of constructed values. I'd consider doing something like `9999 * WEEK`.